File: Medium/2492_Minimum_Score_of_a_Path_Between_Two_Cities.py
class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        
        vis = [False] * (n+1)
        adj = [[] for i in range(n+1)]
        for road in roads:
            adj[road[0]].append((road[1], road[2]))
            adj[road[1]].append((road[0], road[2]))

        def bfs(src):

            ans = int(1e8)
            q = deque()
            vis[src] = True
            q.append(src)

            while q:
                u = q.popleft()
                for pv in adj[u]:
                    if not vis[pv[0]]:
                        vis[pv[0]] = True
                        q.append(pv[0])
                    ans = min(ans, pv[1])
            return ans
        
        ans = int(1e8)
        for i in range(1, n+1):
            if not vis[i] and not vis[1] and not vis[n]:
                ans = min(ans, bfs(i))
            else:
                return ans
        return ans

File: Medium/2492_Minimum_Score_of_a_Path_Between_Two_Cities.cpp
class Solution {
public:

    int bfs(vector<vector<pair<int, int>>> &adj, int src, vector<bool> &vis)
    {
        int ans = 1e8;
        queue<int> q;
        vis[src] = true;
        q.push(src);

        while(!q.empty())
        {
            int u = q.front();
            q.pop();

            for(auto &pv: adj[u])
            {
                if(!vis[pv.first])
                {
                    vis[pv.first] = true;
                    q.push(pv.first);
                }
                ans = min(ans, pv.second);
            }
        }
        return ans;
    }
    int minScore(int n, vector<vector<int>>& roads) {
        
        
        vector<vector<pair<int, int>>> adj(n+1, vector<pair<int, int>> ());
        for(auto &road: roads)
        {
            adj[road[0]].push_back({road[1], road[2]});
            adj[road[1]].push_back({road[0], road[2]});
        }

        int ans = 1e8;
        vector<bool> vis(n+1, false);
        for(int i=1 ; i<=n ; i++)
        {
            if(!vis[i] && !vis[1] && !vis[n])
            {
                ans = min(ans, bfs(adj, i, vis));
            }else
            {
                return ans;
            }
        }
        return ans;
    }
};
